crates/burn-train/src/metric/precision.rs

```diff
@@ -0,0 +1,130 @@
+use super::{
+    classification::{ClassAverageType, ClassificationInput},
+    confusion_stats::ConfusionStats,
+    state::{FormatOptions, NumericMetricState},
+    Metric, MetricEntry, MetricMetadata, Numeric,
+};
+use burn_core::tensor::backend::Backend;
+use core::marker::PhantomData;
+
+/// The precision metric.
+pub struct PrecisionMetric<B: Backend> {
+    state: NumericMetricState,
+    _b: PhantomData<B>,
+    threshold: Option<f64>,
+    class_average: ClassAverageType,
+    top_k: Option<usize>,
+}
+
+#[allow(dead_code)]
+impl<B: Backend> PrecisionMetric<B> {
+    /// Sets the threshold.
+    pub fn with_threshold(mut self, threshold: f64) -> Self {
+        self.threshold = Some(threshold);
+        self.top_k = None;
+        self
+    }
+
+    /// Sets the class average.
+    pub fn with_class_average(mut self, class_average: ClassAverageType) -> Self {
+        self.class_average = class_average;
+        self
+    }
+
+    /// Sets the top k.
+    pub fn with_top_k(mut self, top_k: usize) -> Self {
+        self.top_k = Some(top_k);
+        self.threshold = None;
+        self
+    }
+}
+
+impl<B: Backend> Default for PrecisionMetric<B> {
+    /// Creates a new metric instance with default values.
+    fn default() -> Self {
+        Self {
+            state: NumericMetricState::default(),
+            _b: PhantomData,
+            threshold: Some(0.5),
+            class_average: ClassAverageType::Micro,
+            top_k: None,
+        }
+    }
```

Comment:
Yeah I was avoiding doing that because then if `PrecisionMetric` is initialized as `binary` for example and then `threshold is set if becomes a multiclass/multilable metric and can become confusing in my opinion. We could maybe create the enum `ClassificationType` which could be used to limit  whether we could set threshold or top_k, what do you think?

Reply:
Hmmm yeah that could be annoying.

Maybe a more explicit change could be to have the user-facing metrics to be defined for each type, e.g. `BinaryPrecision`, `MulticlassPrecision` and `MultilabelPrecision`?

Internally it could call common code, but at least we know that the configurations should be valid.

Lmk what you think 🙂

Reply:
Right, that's also how torchmetrics hadles it. I think I agree, I'll make some changes :rocket:

Reply:
Ahh I wasn't aware of that 😄 

Thanks for addressing all of my comments btw 💪

Reply:
Yeah no worries thank you for your patience and effort on this PR :pray: